pkg/devfile/adapters/docker/component/utils.go

```diff
@@ -0,0 +1,167 @@
+	// Create a docker volume to store the project source code
+	volume, err := a.Client.CreateVolume(volumeLabels)
+	if err != nil {
+		return errors.Wrapf(err, "Unable to create project source volume for component %s", componentName)
+	}
+
+	projectVolumeName := volume.Name
+
+	// Loop over each component and start a container for it
+	for _, comp := range supportedComponents {
+		err = a.pullAndStartContainer(componentName, projectVolumeName, comp)
+		if err != nil {
+			return errors.Wrapf(err, "unable to pull and start container %s for component %s", *comp.Alias, componentName)
+		}
+	}
+	glog.V(3).Infof("Successfully created all containers for component %s", componentName)
+
+	return nil
+}
+
+func (a Adapter) updateComponent() (err error) {
+	glog.V(3).Info("The component already exists, attempting to update it")
+	componentName := a.ComponentName
+
+	// Get the project source volume
+	volumeLabels := utils.GetProjectVolumeLabels(componentName)
+	vols, err := a.Client.GetVolumesByLabel(volumeLabels)
+	if err != nil {
+		return errors.Wrapf(err, "Unable to retrieve source volume for component "+componentName)
+	}
+	if len(vols) == 0 {
+		return fmt.Errorf("Unable to find source volume for component %s", componentName)
+	}
+	projectVolumeName := vols[0].Name
+
+	supportedComponents := adaptersCommon.GetSupportedComponents(a.Devfile.Data)
+	if len(supportedComponents) == 0 {
+		return fmt.Errorf("No valid components found in the devfile")
+	}
+
+	for _, comp := range supportedComponents {
+		// Check to see if this component is already running and if so, update it
+		// If component isn't running, re-create it, as it either may be new, or crashed.
+		containers, err := a.Client.GetContainersByComponentAndAlias(componentName, *comp.Alias)
+		if err != nil {
+			return errors.Wrapf(err, "unable to list containers for component %s", componentName)
+		}
+		if len(containers) == 0 {
+			// Container doesn't exist, so need to pull its image (to be safe) and start a new container
+			err = a.pullAndStartContainer(componentName, projectVolumeName, comp)
+			if err != nil {
+				return errors.Wrapf(err, "unable to pull and start container %s for component %s", *comp.Alias, componentName)
+			}
+		} else if len(containers) == 1 {
+			// Container already exists
+			containerID := containers[0].ID
+
+			// Get the associated container config from the container ID
+			containerConfig, err := a.Client.GetContainerConfig(containerID)
+			if err != nil {
+				return errors.Wrapf(err, "unable to get the container config for component %s", componentName)
+			}
+
+			// See if the container needs to be updated
+			if utils.DoesContainerNeedUpdating(comp, containerConfig) {
+				s := log.Spinner("Updating the component " + *comp.Alias)
+				defer s.End(false)
+				// Remove the container
+				err := a.Client.RemoveContainer(containerID)
+				if err != nil {
+					return errors.Wrapf(err, "Unable to remove container %s for component %s", containerID, *comp.Alias)
```

Comment:
do we want to `s.End(false)` here? I checked some of the Kubernetes adapter code, and we dont do it in some of the `if err != nil {` blocks as well, wonder what the UX ramifications are 🤔

Reply:
It's not necessary. The `defer s.End(false)` when we initialize the spinner will make the spinner end with a failure if the function exits at any point before we call `s.End(true)`